lib/optimize/CommonsChunkPlugin.js

```diff
@@ -24,153 +24,304 @@ The available options are:
+			if(!chunk.isInitial()) {
+				for(let block of chunk.blocks) {
+					block.chunks.unshift(commonChunk);
+					commonChunk.addBlock(block);
+				}
+			}
+		}
+	}
+
+	getAsyncChunkOrigin(chunks) {
+		const origins = [];
+		for(let chunk of chunks) {
+			for(let origin of chunk.origins) {
+				const newOrigin = Object.create(origin);
+				newOrigin.reasons = (origin.reasons || []).concat("async commons");
+				origins.push(newOrigin);
+			}
+		}
+		return origins;
+	}
+
 	apply(compiler) {
-		const chunkNames = this.chunkNames;
 		const filenameTemplate = this.filenameTemplate;
-		const minChunks = this.minChunks;
-		const selectedChunks = this.selectedChunks;
 		const asyncOption = this.async;
 		const minSize = this.minSize;
+		const minChunks = this.minChunks;
 		const ident = this.ident;
 		compiler.plugin("this-compilation", (compilation) => {
 			compilation.plugin(["optimize-chunks", "optimize-extracted-chunks"], (chunks) => {
 				// only optimize once
 				if(compilation[ident]) return;
 				compilation[ident] = true;
 
-				let commonChunks;
-				if(!chunkNames && (selectedChunks === false || asyncOption)) {
-					commonChunks = chunks;
-				} else if(Array.isArray(chunkNames) || typeof chunkNames === "string") {
-					commonChunks = [].concat(chunkNames).map((chunkName) => {
-						let chunk = chunks.filter((chunk) => chunk.name === chunkName)[0];
-						if(!chunk) {
-							chunk = compilation.addChunk(chunkName);
-						}
-						return chunk;
-					});
-				} else {
-					throw new Error("Invalid chunkNames argument");
-				}
-				commonChunks.forEach(function processCommonChunk(commonChunk, idx) {
-					let usedChunks;
-					if(Array.isArray(selectedChunks)) {
-						usedChunks = chunks.filter(chunk => chunk !== commonChunk && selectedChunks.indexOf(chunk.name) >= 0);
-					} else if(selectedChunks === false || asyncOption) {
-						usedChunks = (commonChunk.chunks || []).filter((chunk) => {
-							// we can only move modules from this chunk if the "commonChunk" is the only parent
-							return asyncOption || chunk.parents.length === 1;
-						});
-					} else {
-						if(commonChunk.parents.length > 0) {
-							compilation.errors.push(new Error("CommonsChunkPlugin: While running in normal mode it's not allowed to use a non-entry chunk (" + commonChunk.name + ")"));
-							return;
-						}
-						usedChunks = chunks.filter((chunk) => {
-							const found = commonChunks.indexOf(chunk);
-							if(found >= idx) return false;
-							return chunk.hasRuntime();
-						});
+				/**
+				 * Creates a list of common chunks based on the options.
+				 * The list is made up of preexisting or newly created chunks.
+				 * - If chunk has the name as specified in the chunkNames it is put in the list
+				 * - If no chunk with the name as given in chunkNames exists a new chunk is created and added to the list
+				 */
+				const commonChunks = this.getCommonChunks(chunks, compilation, this.chunkNames, this.selectedChunks, this.async);
```

Comment:
You don't need to pass the options to a member function as the function can access the options directly.